x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/DocumentPermissions.java

```diff
@@ -46,58 +48,67 @@
-    // SortedSet because orders are important when they get serialised for request cache key
-    private final SortedSet<BytesReference> queries;
-    private final SortedSet<BytesReference> limitedByQueries;
-    private List<String> evaluatedQueries;
-    private List<String> evaluatedLimitedByQueries;
 
-    private static DocumentPermissions ALLOW_ALL = new DocumentPermissions();
+    @Nullable
+    private final List<Set<BytesReference>> listOfQueries;
+    @Nullable
+    private List<List<String>> listOfEvaluatedQueries;
+
+    private static final DocumentPermissions ALLOW_ALL = new DocumentPermissions();
 
     DocumentPermissions() {
-        this.queries = null;
-        this.limitedByQueries = null;
+        this.listOfQueries = null;
     }
 
     DocumentPermissions(Set<BytesReference> queries) {
-        this(queries, null);
+        if (queries == null || queries.isEmpty()) {
+            throw new IllegalArgumentException("null or empty queries not permitted");
+        }
+        this.listOfQueries = List.of(new TreeSet<>(queries));
     }
 
-    DocumentPermissions(Set<BytesReference> queries, Set<BytesReference> scopedByQueries) {
-        if (queries == null && scopedByQueries == null) {
-            throw new IllegalArgumentException("one of the queries or scoped queries must be provided");
+    DocumentPermissions(List<Set<BytesReference>> listOfQueries) {
+        if (listOfQueries == null || listOfQueries.isEmpty()) {
+            throw new IllegalArgumentException("null or empty list of queries not permitted");
         }
-        this.queries = (queries != null) ? new TreeSet<>(queries) : null;
-        this.limitedByQueries = (scopedByQueries != null) ? new TreeSet<>(scopedByQueries) : null;
+        listOfQueries.forEach(queries -> {
+            if (queries == null || queries.isEmpty()) {
+                throw new IllegalArgumentException("null or empty queries not permitted");
+            }
+        });
+        // SortedSet because orders are important when they get serialised for request cache key
+        this.listOfQueries = listOfQueries.stream()
+            .map(queries -> queries instanceof SortedSet<BytesReference> ? queries : new TreeSet<>(queries))
+            .toList();
     }
 
-    public Set<BytesReference> getQueries() {
-        return queries == null ? null : Set.copyOf(queries);
+    public List<Set<BytesReference>> getListOfQueries() {
+        return listOfQueries;
     }
 
-    public Set<BytesReference> getLimitedByQueries() {
-        return limitedByQueries == null ? null : Set.copyOf(limitedByQueries);
+    public Set<BytesReference> getSingleSetOfQueries() {
+        if (listOfQueries != null && listOfQueries.size() == 1) {
+            return listOfQueries.get(0);
+        } else {
+            assert false;
+            throw new IllegalArgumentException("the list of queries does not have a single member: [" + listOfQueries + "]");
+        }
```

Comment:
A convenient method for when the documentPermissions only has a single set of queries, e.g. normal user roles.

Reply:
Similar question as above ... if this can only execute due to programmer's error can we just assert instead ? (likely negate the if condition as an assertion)

Reply:
I turned the if condition to an assertion. Since this is a new convenient method, it makes sense to do so.